test/test_data.py

```diff
@@ -1,13 +1,92 @@
+from typing import Union, Dict, List, Tuple
 import unittest
+
+import rdflib
+import pytest
+
 from pyontutils.core import OntResIri
 
+Semantic = Union[rdflib.URIRef, rdflib.Literal, rdflib.BNode]
+
 
 class TestCurationExportTtl(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
         cls.ori = OntResIri('https://cassava.ucsd.edu/sparc/exports/curation-export.ttl')
         cls.graph = cls.ori.graph
+        cls.spaql_templates = SparqlQueryTemplates(cls.graph.namespace_manager)
 
-    def test(self):
+    def test_dataset_group(self):
         """ sparql queries here """
-        self.graph
+        subj = 'https://api.blackfynn.io/datasets/N:dataset:c2a014b8-2c15-4269-b10a-3345420e3d56/subjects/53'
+        query = self.spaql_templates.dataset_group(subj)
+        assert len(list(self.graph.query(query))) > 0
+
+    def test_related_datasets(self):
+        subj = 'dataset:bec4d335-9377-4863-9017-ecd01170f354'
+        query = self.spaql_templates.related_datasets(subj)
+        assert len(list(self.graph.query(query))) > 0
+
+
+class SparqlQueryTemplates:
+    """ Creates SPARQL query templates. """
+
+    def __init__(self, nsm=None):
+        self.nsm = nsm if nsm else rdflib.Graph().namespace_manager
+        self.prefixes = {p:ns for p, ns in self.nsm.namespaces() if p}
+
+    def sparql_uri(self, uri_or_curie: Union[rdflib.URIRef, str]) -> str:
+        """ Converts IRIs and curies to a usable format for SPARQL queries. """
+        uri_or_curie = str(uri_or_curie)
+        if uri_or_curie.startswith('http') or isinstance(uri_or_curie, rdflib.URIRef):
+            return '<'+str(uri_or_curie)+'>'
+        try:
+            # Expand as a trick to see if the curie prefix provided exists.
+            return '<'+str(rdflib.util.from_n3(uri_or_curie, nsm=self.nsm))+'>'
+        except:
+            ValueError(f'Invalid prefix used {uri_or_curie}')
+
+    def prefixes_used(self, query: str) -> str:
+        """ Parses SPARQL query to see what curie prefix are used & returns SPARQL PREFIX block. """
+        prefixes = set()
+        for e in query.split():
+            if ':' in e:
+                prefix = e.split(':', 1)[0]
+                if self.prefixes.get(prefix):
+                    prefixes.add(f'PREFIX {prefix}: <{self.prefixes[prefix]}>\n')
+        return '\n'.join(prefixes)
+
+    def dataset_group(self, subject: Union[rdflib.URIRef, str]) -> List[Tuple[Semantic, Semantic, Semantic]]:
+        """ Get all subject groups and dataset associated with subject input.
+
+        :returns: list of tuples containing: subject, subjects group, and subjects dataset.
+        """
+        subject = self.sparql_uri(subject)
+        query = """
+            SELECT ?subj ?group ?dataset
+            WHERE {
+                %s TEMP:hasAssignedGroup ?group .
+                ?subj TEMP:hasAssignedGroup ?group .
+                ?subj TEMP:hasDerivedInformationAsParticipant ?dataset .
+            }
+        """ % (subject)
```

Comment:
While SPARQL injection isn't quite as big a concern as SQL injection, it is probably best to use the prepared query [1] functionality in cases like this (including the cases here). Also for consistency all string formatting in this repo should use fstrings instead of the modulus operator.

1. https://rdflib.readthedocs.io/en/stable/intro_to_sparql.html#prepared-queries

Reply:
I thought fstring wouldn't work, but it turns out double brackets allow in-string brackets to exist. I don't know why it took me so long to find that out.